### PR TITLE
Add kubeconfig link to ~/.kube/config

### PR DESCRIPTION
**Pull request by erikwilson**
For rancher/k3s#1126

### CODE REVIEW

**Comment by ibuildthecloud** [comment]
LGTM

**Comment by nicdesousa** [comment]
@erikwilson Was this PR delayed due to `/etc/rancher/k3s/k3s.yaml` perms?

```bash
$ file $(readlink -f ~/.kube/config)
/etc/rancher/k3s/k3s.yaml: regular file, no read permission
...
$ getfacl $(readlink -f ~/.kube/config)
# file: etc/rancher/k3s/k3s.yaml
# owner: root
# group: root
user::rw-
group::---
other::---
```

Unless there are better solutions:
The installation script can update the `/etc/rancher/k3s/k3s.yaml` file ACL to enable read-only access for the user account group.
>Note: This requires the `acl` package to be installed... A test to check if the `setfacl` binary is available should be sufficient.
```bash
setfacl -m g:$(id -g):r /etc/rancher/k3s/k3s.yaml
```
Result (the user account group is `nic` in my case):
```bash
$ getfacl $(readlink -f ~/.kube/config)               
# file: etc/rancher/k3s/k3s.yaml
# owner: root
# group: root
user::rw-
group::---
group:nic:r--
mask::r--
other::---
...
$ kubectl cluster-info
Kubernetes master is running at https://127.0.0.1:6443
...
```

File: install.sh [hunk 1]
```diff
@@ -123,6 +127,50 @@ escape_dq() {
     printf '%s' "$@" | sed -e 's/"/\\"/g'
 }
 
+# --- check for rootless mode ---
+is_rootless() {
+    while [ $# != 0 ]; do
```
Comment (joakimr-axis) [inline install.sh]: Better use -ne (numerical comparison) rather than ta string comparison.

[ $# -ne 0 ]

File: install.sh [hunk 2]
```diff
@@ -123,6 +127,50 @@ escape_dq() {
...
+        return
+    fi
+
+    if is_rootless "$@"; then
+        KUBECONFIG=$HOME/.kube/k3s.yaml
+    else
+        KUBECONFIG=/etc/rancher/k3s/k3s.yaml
+    fi
+
+    while [ $# != 0 ]; do
```
Comment (joakimr-axis) [inline install.sh]: [ $# -ne 0 ]

File: install.sh [hunk 3]
```diff
@@ -733,6 +783,37 @@ service_enable_and_start() {
...
+        for _ in $(seq 1 20); do
+            [ -f "${KUBECONFIG}" ] && break
+            sleep 1
+        done
+        [ -f "${KUBECONFIG}" ] || warn "Kubeconfig does not yet exist at ${KUBECONFIG}"
+    fi
+
+    if [ -L ${DEFAULT_CONFIG} ]; then
+        CONFIG_LINK=$(readlink -f ${DEFAULT_CONFIG} || true)
+        if [ "${CONFIG_LINK}" == "${KUBECONFIG}" ] || echo ${CONFIG_LINK} | grep -vq k3s; then
```
Comment (joakimr-axis) [inline install.sh]: Better use = than == since the POSIX single = is what the existing k3s scripts (almost) consistently use.

File: install.sh [hunk 4]
```diff
@@ -733,6 +783,37 @@ service_enable_and_start() {
...
+            info "Skipping kubeconfig symlink, ~/.kube/config link already exists to ${CONFIG_LINK}"
+            return
+        fi
+    elif [ -f ${DEFAULT_CONFIG} ]; then
```
Comment (joakimr-axis) [inline install.sh]: "${DEFAULT_CONFIG}"

File: install.sh [hunk 5]
```diff
@@ -733,6 +783,37 @@ service_enable_and_start() {
...
+        info "Skipping kubeconfig symlink, ~/.kube/config file already exists"
+        return
+    fi
+
+    if [ ! -d $(dirname ${DEFAULT_CONFIG}) ]; then
```
Comment (joakimr-axis) [inline install.sh]: Add (double) quotes around $(dirname ${DEFAULT_CONFIG]).
Furthermore, unnecessary call to dirname when the shell already provides this for you:
${DEFAULT_CONFIG%/*}

File: install.sh [hunk 6]
```diff
@@ -733,6 +783,37 @@ service_enable_and_start() {
...
+        mkdir -p $(dirname ${DEFAULT_CONFIG})
```
Comment (joakimr-axis) [inline install.sh]: See comment about quotes and dirname above.

File: install.sh [hunk 7]
```diff
@@ -733,6 +783,37 @@ service_enable_and_start() {
...
+    fi
+
+    ln -sf ${KUBECONFIG} ${DEFAULT_CONFIG}
```
Comment (joakimr-axis) [inline install.sh]: Without quotes here, eventual whitespaces in the file names/paths will break this.

**Comment by xiaods** [comment]
please resolve the conflict file. @erikwilson

**Comment by dereknola** [comment]
Closing this as Out-Of-Date.

**Pull request closed without merging**